### PR TITLE
Added run-checks cli command

### PR DESCRIPTION
**Pull request by maallen**
New cli command that executes checks against any new strings that have been added. Checks are run in parallel and can be configurable on a checker level to be a hard-failure i.e. return an error code to the command line.

### CODE REVIEW

File: cli/src/main/java/com/box/l10n/mojito/cli/command/CheckerCommand.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        return regexSet;
+    }
+
+    private List<CliChecker> generateChecks(List<AssetExtractionDiff> assetExtractionDiffs) {
+        List<CliChecker> cliCheckers = new ArrayList<>();
+        String[] checks = checkerList.split(",");
+        CliCheckerOptions options = generateCheckerOptions();
+        Stream<CliCheckerType> checkerTypes = Arrays.stream(CliCheckerType.values());
+        for(String check : checks) {
+            if(checkerTypes.anyMatch(c -> c.name().equals(check))){
```
Comment (aurambaj) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/CheckerCommand.java:166]: having something like `Stream.of(checks).map(c -> createInstance...)`? Not sure why we have a double check with `checkerTypes` 

`checkerTypes.anyMatch()` calling twice `anyMatch()` on a stream will fail no?

File: cli/src/main/java/com/box/l10n/mojito/cli/command/checks/CliCheckerExecutor.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public List<String> executeChecks() {
+        try {
+            List<Future<CliCheckResult>> futures = startChecks(executorService);
+            return analyzeResults(retrieveCheckResults(futures));
+        } catch (InterruptedException | ExecutionException e) {
+            throw new CommandException("Failed to retrieve results from check executor: " + e.getMessage(), e);
+        } finally {
+            executorService.shutdown();
```
Comment (aurambaj) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/checks/CliCheckerExecutor.java:36]: what happens if this is called twice? 

else spring a bunch of facilities to run async task that may help hide some of this setup (executor service, etc)

File: cli/src/main/java/com/box/l10n/mojito/cli/command/CheckerCommand.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                .diffExtractionName(extractionDiffName)
+                .baseExtractorPaths(baseExtractionPaths)
+                .currentExtractorPaths(currentExtractionPaths)
+                .build();
+
+        List<AssetExtractionDiff> assetExtractionDiffs;
+
+        try {
+            if (extractionDiffService.hasAddedTextUnits(extractionDiffPaths)) {
+                assetExtractionDiffs = extractionDiffService.computeAssetExtractionDiffs(extractionDiffPaths);
```
[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/CheckerCommand.java:98]: Need to clean that method up, it was done as a quick way of getting the asset extraction diffs in memory and then I never put in a TODO so missed it when I came back.

File: pom.xml [hunk 4]
```diff
@@ -64,6 +64,16 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
+            <exclusions>
```
[the remark this answers is not in the document]
Comment (maallen) [reply pom.xml:67]: I don't think it does as the dependency tree shows its pulling in junit 4.13.

`[INFO] +- org.springframework.boot:spring-boot-starter-test:jar:2.3.4.RELEASE:test
[INFO] |  +- org.springframework.boot:spring-boot-test:jar:2.3.4.RELEASE:test
[INFO] |  +- org.springframework.boot:spring-boot-test-autoconfigure:jar:2.3.4.RELEASE:test
...
[INFO] |  +- org.junit.vintage:junit-vintage-engine:jar:5.6.2:test
[INFO] |  |  +- org.apiguardian:apiguardian-api:jar:1.1.0:test
[INFO] |  |  +- org.junit.platform:junit-platform-engine:jar:1.6.2:test
[INFO] |  |  \- junit:junit:jar:4.13:compile
...`

However when I include the hunspell dependency the cli unit tests failed to run with the below error:

`Nov 03, 2021 5:26:17 PM org.junit.platform.launcher.core.DefaultLauncher handleThrowable
WARNING: TestEngine with ID 'junit-vintage' failed to discover tests
org.junit.platform.commons.JUnitException: Failed to parse version of junit:junit: 3.8.1
        at org.junit.vintage.engine.JUnit4VersionCheck.parseVersion(JUnit4VersionCheck.java:54)
        at org.junit.vintage.engine.JUnit4VersionCheck.checkSupported(JUnit4VersionCheck.java:37)
        at org.junit.vintage.engine.JUnit4VersionCheck.checkSupported(JUnit4VersionCheck.java:32)
        at org.junit.vintage.engine.VintageTestEngine.discover(VintageTestEngine.java:62)
        at org.junit.platform.launcher.core.DefaultLauncher.discoverEngineRoot(DefaultLauncher.java:168)
        at org.junit.platform.launcher.core.DefaultLauncher.discoverRoot(DefaultLauncher.java:155)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:150)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:124)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)`
        
Adding the exclusions allowed the tests to run correctly, the hunspell dependency itself doesn't appear to pull in any junit dependencies so I'm unsure why I was seeing the above error.

[INFO] +- com.gitlab.dumonts:hunspell:jar:1.1.0:compile
[INFO] |  \- com.nativelibs4java:bridj:jar:0.7.0:compile
[INFO] |     \- com.google.android.tools:dx:jar:1.7:compile
[INFO] +- org.springframework.boot:spring-boot-starter-logging:jar:2.3.4.RELEASE:compile

File: cli/src/main/java/com/box/l10n/mojito/cli/command/checks/CliCheckerExecutor.java (hunk 2 above)
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/checks/CliCheckerExecutor.java:36]: I'll move the creation of the thread pool to be inside the method, in that case a threadpool is created and shutdown on each call to executeChecks.

**Comment by maallen** [comment]
Split review into different PRs: https://github.com/box/mojito/pull/738 & https://github.com/box/mojito/pull/742

**Pull request closed without merging**